File: src/standard_ai.c
/*
The MIT License (MIT)

Copyright (c) 2014 Master.G

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include "standard_ai.h"
#include "game.h"

int StandardAI_GetReady(void *p, void *game) {
  player_t *player = (player_t *) p;

  CardArray_Sort(&player->cards, NULL);
  CardArray_Copy(&player->record, &player->cards);
#if (PRINT_GAME_LOG == 1)
  CardArray_Print(&player->record);
#endif /* ifdef PRINT_GAME_LOG */
  // FIXME
  player->handlist = HandList_StandardAnalyze(&player->cards);

  HandList_Print(player->handlist);

  return 0;
}

int StandardAI_Bid(void *p, void *g) {
  int shouldbid = 0;
  int handlistlen = 0;
  game_t *game = (game_t *) g;
  player_t *player = (player_t *) p;

  CardArray_Sort(&player->cards, NULL);
  player->handlist = HandList_StandardAnalyze(&player->cards);
  handlistlen = rk_list_count(player->handlist);
  rk_list_clear_destroy(player->handlist);

  if (handlistlen > 9) {
    shouldbid = 0;
  } else if ((handlistlen < 9) && (handlistlen > 3)) {
    shouldbid = 1;
  } else if ((handlistlen <= 3) && (handlistlen > 2)) {
    shouldbid = 2;
  } else if (handlistlen <= 2) {
    shouldbid = 3;
  }

  if (shouldbid > game->bid)
    return shouldbid;
  else
    return 0;
}

int StandardAI_Play(void *p, void *game) {
  int countpair = 0;
  int countsolo = 0;
  int need = 0;
  int searchprimal = 0;
  int kicker = 0;
  player_t *player = (player_t *) p;
  hand_t *node = NULL;
  rk_list_node_t *temp = NULL;

  hand_t *hand = &((game_t *) game)->lastHand;

  do {
    /* empty hands */
    if (player->handlist == NULL) {
      hand->type = 0;
      break;
    }

    /* last hand */
    if (rk_list_count(player->handlist) == 1) {
      Hand_Copy(hand, (hand_t *) player->handlist->first->payload);
      HandList_Remove(player->handlist, player->handlist->first->payload);
      break;
    }

    /* try to find longest hand combination */
    node = HandList_Find(player->handlist,
                         Hand_Format(HAND_PRIMAL_TRIO,
                                     HAND_KICKER_NONE,
                                     HAND_CHAIN));

    if (node != NULL) {
      /* copy hand*/
      Hand_Copy(hand, node);

      /* how many kickers do we need */
      need = node->cards.length / 3;

      /* remove hand node from hand list */
      HandList_Remove(player->handlist, node);

      /* count solo and pair number */
      temp = player->handlist->first;
      countpair = 0;
      countsolo = 0;

      while (temp->next != NULL) {
        if (HandList_GetHand(temp)->type == HAND_PRIMAL_PAIR)
          countpair++;
        else if (HandList_GetHand(temp)->type == HAND_PRIMAL_SOLO)
          countsolo++;

        temp = temp->next;
      }

      /* trio-pair-chain then trio-solo-chain */
      if ((countsolo < need) && (countpair < need)) {
        break;
      } else if (countpair >= need) {
        searchprimal = HAND_PRIMAL_PAIR;
        kicker = HAND_KICKER_PAIR;
      } else {
        searchprimal = HAND_PRIMAL_SOLO;
        kicker = HAND_KICKER_SOLO;
      }

      /* detach pairs from list */
      temp = player->handlist->first;

      while (need > 0 && temp != NULL) {
        if (HandList_GetHand(temp)->type ==
          Hand_Format(searchprimal, HAND_KICKER_NONE, HAND_CHAINLESS)) {
          /* copy cards */
          CardArray_Concat(&hand->cards, &HandList_GetHand(temp)->cards);
          HandList_Remove(player->handlist, temp->payload);
          temp = player->handlist->first;
          need--;
        } else {
          temp = temp->next;
        }
      }

      Hand_SetKicker(hand->type, kicker);
      break;
    }

    /* pair chain */
    node = HandList_Find(player->handlist,
                    Hand_Format(HAND_PRIMAL_PAIR,
                                HAND_KICKER_NONE,
                                HAND_CHAIN));

    /* solo chain */
    if (node == NULL)
      node = HandList_Find(player->handlist,
                           Hand_Format(HAND_PRIMAL_SOLO, HAND_KICKER_NONE,
                                       HAND_CHAIN));

    if (node != NULL) {
      Hand_Copy(hand, node);
      HandList_Remove(player->handlist, node);
      break;
    }

    /* trio */
    node = HandList_Find(player->handlist,
                         Hand_Format(HAND_PRIMAL_TRIO, HAND_KICKER_NONE,
                                     HAND_CHAINLESS));

    if ((node != NULL) &&
      (CARD_RANK(node->cards.cards[0]) != CARD_RANK_2)) {
      Hand_Copy(hand, node);
      HandList_Remove(player->handlist, node);

      /* pair */
      node = HandList_Find(player->handlist,
                           Hand_Format(HAND_PRIMAL_PAIR, HAND_KICKER_NONE,
                                       HAND_CHAINLESS));

      if ((node != NULL) &&
        (CARD_RANK(node->cards.cards[0]) != CARD_RANK_2)) {
        kicker = HAND_KICKER_PAIR;
      } else {
        /* solo */
        node = HandList_Find(player->handlist,
                             Hand_Format(HAND_PRIMAL_SOLO, HAND_KICKER_NONE,
                                         HAND_CHAINLESS));

        if ((node != NULL) &&
          (CARD_RANK(node->cards.cards[0]) < CARD_RANK_2))
          kicker = HAND_KICKER_SOLO;
      }

      if (node != NULL) {
        CardArray_Concat(&hand->cards, &node->cards);
        Hand_SetKicker(hand->type, kicker);
        HandList_Remove(player->handlist, node);
        break;
      }

      /* no solo nor pair, return with trio */
      break;
    }

    /* pair */
    node = HandList_Find(player->handlist,
                         Hand_Format(HAND_PRIMAL_PAIR, HAND_KICKER_NONE,
                                     HAND_CHAINLESS));

    if ((node != NULL) &&
      (CARD_RANK(node->cards.cards[0]) != CARD_RANK_2)) {
      Hand_Copy(hand, node);
      HandList_Remove(player->handlist, node);
      break;
    }

    /* just play */
    node = player->handlist->first->payload;
    Hand_Copy(hand, node);
    HandList_Remove(player->handlist, node);
  } while (0);

  CardArray_Subtract(&player->cards, &hand->cards);

  return 0;
}

int StandardAI_Beat(void *p, void *g) {
  /*
   * HandList_SearchBeats can search for beat in loop mode
   * but we just simply find a beat here
   */
  int canbeat = 0;
  int i = 0;
  hand_t *tobeat;
  hand_t beat;
  player_t *player = (player_t *) p;
  player_t *prevplayer = NULL;
  player_t *teammate = NULL;
  player_t *landlord = NULL;
  game_t *game = (game_t *) g;

  Hand_Clear(&beat);

  tobeat = &((game_t *) game)->lastHand;

  canbeat = HandList_BestBeat(&player->cards, tobeat, &beat, NULL);

  /*
     canbeat = HandList_SearchBeat(&player->cards, tobeat, &beat);
   */

  /* peasant cooperation */
  prevplayer = &game->players[game->lastplay];

  if (canbeat &&
    (player->identity == PlayerIdentity_Peasant) &&
    (prevplayer->identity == PlayerIdentity_Peasant)) {
    /* find teammate and landlord */
    for (i = 0; i < GAME_PLAYERS; i++) {
      if (game->players[i].identity == PlayerIdentity_Landlord)
        landlord = &game->players[i];

      if ((game->players[i].identity == PlayerIdentity_Peasant) &&
        (game->players[i].seatId !=
          player->seatId))
        teammate = &game->players[i];
    }

    /* don't bomb/nuke teammate */
    if (canbeat &&
      (((beat.type ==
        Hand_Format(HAND_PRIMAL_BOMB, HAND_KICKER_NONE, HAND_CHAINLESS)) ||
        (beat.type ==
          Hand_Format(HAND_PRIMAL_NUKE, HAND_KICKER_NONE,
                      HAND_CHAINLESS)))))
      canbeat = 0;

    if (canbeat && (teammate->cards.length < player->cards.length)) canbeat = 0;
  }

  if (canbeat) {
    CardArray_Subtract(&player->cards, &beat.cards);
    rk_list_clear_destroy(player->handlist);
    player->handlist = HandList_StandardAnalyze(&player->cards);
    Hand_Copy(tobeat, &beat);
  }

  return canbeat;
}

File: lualandlord/standard_ai.lua
require "lualandlord/hand"

function ll.StandardAI_GetReady(player, game)
	ll.CardArray_Sort(player.cards);
	player.record = ll.CardArray_Copy(player.cards);
	player.handlist = ll.HandList_StandardAnalyze(player.cards);

	return 0;
end

function ll.StandardAI_Bid(player, game)
	local shouldbid = 0;
	local handlistlen = 0;
	ll.CardArray_Sort(player.cards);
	player.handlist = ll.HandList_StandardAnalyze(player.cards);
	handlistlen = #player.handlist;

	if handlistlen > 12 then
		shouldbid = 0;
	elseif handlistlen == 12 then
		shouldbid = 1;
	elseif handlistlen == 11 then
		shouldbid = 2;
	elseif handlistlen < 11 then
		shouldbid = 3;
	end

	if shouldbid > game.bid then
		return shouldbid;
	else
		return 0;
	end
end

function ll.StandardAI_Play(player, game)
	local need = 0;
	local searchprimal = 0;
	local kicker = 0;

	local hand = game.lastHand;

	repeat
		-- empty hands
		if #player.handlist == 0 then
			hand.type = 0;
			break;
		end

		-- last hand
		if #player.handlist == 1 then
			local last = player.handlist[1];
			hand.type = last.type;
			hand.cards = ll.CardArray_Copy(last.cards);
			player.handlist = {};
			break;
		end

		-- try to find longest hand combination
		local node = ll.HandList_Find(player.handlist, ll.Hand_Format(ll.HAND_PRIMAL_TRIO, ll.HAND_KICKER_NONE, ll.HAND_CHAIN));
		if node ~= nil then
			-- copy hand
			hand.type = node.type;
			hand.cards = ll.CardArray_Copy(node.cards);

			-- how many kickers do we need
			need = node.cards.length / 3;

			-- remove hand node from list
			ll.HandList_Remove(player.handlist, node);

			-- count solo and pair number
			local countpair = 0;
			local countsolo = 0;
			for k, v in pairs(player.handlist) do
				if v.type == ll.HAND_PRIMAL_PAIR then
					countpair = countpair + 1;
				elseif v.type == ll.HAND_PRIMAL_SOLO then
					countsolo = countsolo + 1;
				end
			end

			-- trio-pair-chain then trio-solo-chain
			if countsolo < need and countpair < need then
				break;
			elseif countpair >= need then
				searchprimal = ll.HAND_PRIMAL_PAIR;
				kicker = ll.HAND_KICKER_PAIR;
			else
				searchprimal = ll.HAND_PRIMAL_SOLO;
				kicker = ll.HAND_KICKER_SOLO;
			end

			-- detach pairs from list
			for k, v in pairs(player.handlist) do
				if need <= 0 then
					break;
				end

				local rest = {};
				if v.type == searchprimal then
					-- copy cards
					ll.CardArray_Concate(hand.cards, v.cards);
					need = need - 1;
				else
					table.insert(rest, v);
				end
			end

			player.handlist = rest;

			hand.type = ll.Hand_SetKicker(hand.type, kicker);
			break;
		end

		-- pair chain
		node = ll.HandList_Find(player.handlist, ll.Hand_Format(ll.HAND_PRIMAL_PAIR, ll.HAND_KICKER_NONE, ll.HAND_CHAIN));
		-- solo chain
		if node == nil then
			node = ll.HandList_Find(player.handlist, ll.Hand_Format(ll.HAND_PRIMAL_SOLO, ll.HAND_KICKER_NONE, ll.HAND_CHAIN));
		end

		if node ~= nil then
			hand.type = node.type;
			hand.cards = ll.CardArray_Copy(node.cards);
			ll.HandList_Remove(player.handlist, node);
			break;
		end

		-- trio
		node = ll.HandList_Find(player.handlist, ll.HAND_PRIMAL_TRIO);
		if node ~= nil and ll.Card_Rank(node.cards.cards[1]) ~= ll.CARD_RANK_2 then
			hand.type = node.type;
			hand.cards = ll.CardArray_Copy(node.cards);

			ll.HandList_Remove(player.handlist, node);

			-- pair
			node = ll.HandList_Find(player.handlist, ll.HAND_PRIMAL_PAIR);
			if node ~= nil and ll.Card_Rank(node.cards.cards[1]) ~= ll.CARD_RANK_2 then
				kicker = ll.HAND_KICKER_PAIR;
			else
				-- solo
				node = ll.HandList_Find(player.handlist, ll.HAND_PRIMAL_SOLO);
				if node ~= nil and ll.Card_Rank(node.cards.cards[1]) ~= ll.CARD_RANK_2 then
					kicker = ll.HAND_KICKER_SOLO;
				end
			end

			if node ~= nil then
				ll.CardArray_Concate(hand.cards, node.cards);
				hand.type = ll.Hand_SetKicker(hand.type, kicker);
				ll.HandList_Remove(player.handlist, node);
				break;
			end
		end

		-- pair
		node = ll.HandList_Find(player.handlist, ll.HAND_PRIMAL_PAIR);
		if node ~= nil and ll.Card_Rank(node.cards.cards[1]) ~= ll.CARD_RANK_2 then
			hand.type = node.type;
			hand.cards = ll.CardArray_Copy(node.cards);
			ll.HandList_Remove(player.handlist, node);
			break;
		end

		-- just play
		node = player.handlist[1];
		hand.type = node.type;
		hand.cards = ll.CardArray_Copy(node.cards);
		ll.HandList_Remove(player.handlist, node);

	until true;

	ll.CardArray_Subtract(player.cards, hand.cards);

	return 0;
end

function ll.StandardAI_Beat(player, game)
	-- HandList_SearchBeats can search for beat in loop mode
	-- but we just simply find a beat here

	local canbeat = false;
	local i = 0;
	local tobeat = nil;
	local beat = ll.Hand_Create();
	local prevplayer = nil;
	local teammate = nil;
	local landlord = nil;

	tobeat = game.lastHand;

	canbeat = ll.HandList_SearchBeat(player.cards, tobeat, beat);

	-- peasant teamwork
	prevplayer = game.players[game.lastplay];
	if canbeat and 
		(player.identity == ll.PlayerIdentity_Peasant) and
		(prevplayer.identity == ll.PlayerIdentity_Peasant) then

		-- find teammate and landlord
		for k, v in pairs(game.players) do
			if (v.identity == ll.PlayerIdentity_Landlord) then
				landlord = v;
			end

			if (v.identity == ll.PlayerIdentity_Peasant) and (v ~= player) then
				teammate = v;
			end
		end

		-- don't bomb/nuke teammate
		if canbeat and ((beat.type == ll.HAND_PRIMAL_NUKE) or (beat.type == ll.HAND_PRIMAL_BOMB)) then
			canbeat = false;
		end

		if canbeat and teammate.cards.length < player.cards.length then
			canbeat = false;
		end
	end

	if canbeat then
		ll.CardArray_Subtract(player.cards, beat.cards);
		player.handlist = ll.HandList_StandardAnalyze(player.cards);
		tobeat.type = beat.type;
		tobeat.cards = ll.CardArray_Copy(beat.cards);
	end

	return canbeat;
end
